hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java

```diff
@@ -250,11 +250,11 @@ public void setLastCleanupTimeCopy(long lastCleanupTimeCopy) {
      *                     cleanup is necessary or not
      */
     final void cleanUpOnMigration(int replicaIndex) {
-        mapService.getMapServiceContext().getMapContainers().keySet()
+        mapService.getMapServiceContext().getMapContainers().entrySet()
                 .stream()
-                .filter(mapName -> replicaIndex == -1
-                        || lesserBackupMapsThen(replicaIndex).test(getRecordStore(mapName)))
```

Comment:
This is running after the record store was removed from `PartitionContainer#maps` -> new `RecordStore` instance.